runtime/src/iree/modules/hal/inline/module.c

```diff
@@ -0,0 +1,609 @@
+// to track lifetime combined with a custom allocator to handle
+// cleaning up this wrapper when the VM buffer is no longer referenced.
+//
+// Since the HAL buffer is providing the storage and the VM buffer is just
+// pointing into it the critical thing this wrapper does is ensure the HAL
+// buffer always outlives the VM buffer.
+//
+// NOTE: this is allocated each storage query! The assumption is that the
+// returned buffer is long-lived (at least per-invocation). This is primarily
+// used to get the backing storage of a !hal.buffer that a user passes into an
+// invocation and the compiler should CSE such queries. Since users can provide
+// their own allocators they can decide if they want to pool small allocations
+// to bypass the system allocator. If we wanted to in here we could have a small
+// free list we maintained for this purpose at the cost of fixed memory
+// consumption. Note that the key requirement is that the returned VM buffer
+// may outlive the module so we can't use an arena that has module lifetime.
+typedef struct iree_hal_inline_storage_buffer_t {
+  // Allocator used to allocate this storage buffer.
+  iree_allocator_t host_allocator;
+  // HAL buffer backing this storage buffer.
+  // Retained for the lifetime of this instance so that the
+  // wrapped vm_buffer is always valid.
+  iree_hal_buffer_t* hal_buffer;
+  // Scoped mapping into the buffer. We could make it persistent but because
+  // we can trivially scope things having this extra information is cheap and
+  // useful for debugging.
+  iree_hal_buffer_mapping_t mapping;
+  // Inline initialized VM buffer wrapping the hal_buffer storage.
+  // This directly references the memory of the HAL buffer.
+  // The buffer has a custom allocator that calls back into this
+  // struct to deallocate the wrapper.
+  iree_vm_buffer_t vm_buffer;
+} iree_hal_inline_storage_buffer_t;
+
+static void iree_hal_inline_storage_buffer_destroy(
+    iree_hal_inline_storage_buffer_t* storage);
+
+static iree_status_t iree_hal_inline_storage_buffer_ctl(
+    void* self, iree_allocator_command_t command, const void* params,
+    void** inout_ptr) {
+  if (command != IREE_ALLOCATOR_COMMAND_FREE) {
+    return iree_make_status(
+        IREE_STATUS_FAILED_PRECONDITION,
+        "allocator can only be used for dropping the wrapper buffer");
+  }
+  iree_hal_inline_storage_buffer_t* storage =
+      (iree_hal_inline_storage_buffer_t*)self;
+  iree_hal_inline_storage_buffer_destroy(storage);
+  return iree_ok_status();
+}
+
+// Creates a VM buffer wrapper that directly references HAL buffer storage.
+// The returned |out_vm_buffer| lifetime will extend the HAL buffer lifetime.
+static iree_status_t iree_hal_inline_storage_buffer_create(
+    iree_hal_buffer_t* hal_buffer, iree_allocator_t host_allocator,
+    iree_vm_buffer_t** out_vm_buffer) {
+  IREE_ASSERT_ARGUMENT(hal_buffer);
+  IREE_ASSERT_ARGUMENT(out_vm_buffer);
+  *out_vm_buffer = NULL;
+  IREE_TRACE_ZONE_BEGIN(z0);
```

Comment:
How do traces between the inline HAL and the "full" HAL compare? I'm wondering both qualitatively (is Tracy "compare traces" able to line up matching parts of the program?) and quantitatively (how much does the inline HAL reduce overheads?).

Reply:
Cool. I'm curious to see how far we can take the emitc path (may need to tweak generated code patterns so the C++ compiler can fold more away)